### PR TITLE
Add "Supersingular Isogeny Diffie-Hellman 101"

### PR DESCRIPTION
**Pull request by lvh**
In a comment on the pull request, include the following information and check off the list.

 - [x] Mini bio
  - Soft capped to 75 words
 - [x] Desired date of publishing
  - Make sure this matches the `date` in the front matter
 - [x] Social media contact info
 - [x] Redaction: (yes|no)
  - Allow redaction, editing or structure changes by technical writer staff
 - [x] Comments: (yes|no)
  - If you're willing to answer comments about your post, set `comments: true`.
  - Make sure to subscribe to your post comments by clicking on the **Subscribe** link at the very bottom of the post _once_ it's published.

@catlook Please review.

### CODE REVIEW

**Comment by lvh** [comment]
* Mini bio:

> lvh is a cryptographer and distributed systems engineer, working as Principal Engineer at Rackspace Managed Security. His beard solves discrete logs in fields of small characteristic.

* Desired date of publishing: Today, May 12th
* Social media contact info: lvh everywhere (@lvh on twitter, lvh on keybase, lvh on hacker news, lvh on reddit, lvh@rackspace.com, _@lvh.io...)
* Redaction: possibly, although inclined to suggest not; this is a highly in depth post on a niche topic with a decent amount of vocabulary that can't be changed without changing the meaning
* Comments: I'd rather get e-mail :)

File: _posts/2016-05-12-supersingular-isogeny-diffie-hellman-101.md [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+date: 2016-05-12 12:15
+comments: false
+author: lvh
+published: true
+categories:
+    - security
+    - cryptography
+---
+
+Craig Costello, Patrick Longa and Michael Naehrig, three cryptographers at
```
Comment (catlook) [inline _posts/2016-05-12-supersingular-isogeny-diffie-hellman-101.md]: ... Longa, and Michael ...

File: _posts/2016-05-12-supersingular-isogeny-diffie-hellman-101.md [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+Microsoft Research, recently published a [paper][costello] on supersingular
+isogeny Diffie-Hellman. This paper garnered a lot of interest in the security
+community and even made it to the front page of Hacker News. Most of the
+discussion around it seemed to be how no-one understands isogenies, even
```
Comment (catlook) [inline _posts/2016-05-12-supersingular-isogeny-diffie-hellman-101.md]: ... how no one understands  ...

File: _posts/2016-05-12-supersingular-isogeny-diffie-hellman-101.md [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+within cryptography-literate communities. This article aims to give you a
+high-level understanding of what this cryptosystem is and why it works.
+
+This post assumes that you already know how Diffie-Hellman works in the
+abstract, and that you know elliptic curves are a mathematical construct that
```
Comment (catlook) [inline _posts/2016-05-12-supersingular-isogeny-diffie-hellman-101.md]: ... abstract and that ...

File: _posts/2016-05-12-supersingular-isogeny-diffie-hellman-101.md [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+"post-quantum" cryptosystems. Those are cryptosystems that will remain secure
+even if the attacker has access to a large quantum computer. This has nothing
+to do with quantum cryptography (for example, quantum key distribution)
+beyond their shared quantum mechanical underpinning.
+
+**Why should I care about quantum computers?**
+
+General quantum computers are not useful as general-purpose computing devices,
+but they can solve some problems much faster than classical
+computers. Classical computers can emulate quantum computers, but only with
```
Comment (catlook) [inline _posts/2016-05-12-supersingular-isogeny-diffie-hellman-101.md]: ... computers, but only ...

File: _posts/2016-05-12-supersingular-isogeny-diffie-hellman-101.md [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+ECDH and ECDSA). To quantify that, consider the following table:
+
+![quantum computer attack cost versus classical][qcost]
+[qcost]: /img/post-quantum/quantum-computer-relative-cost.png
+
+In this table, n refers to the modulus size for RSA, and the field size for
+ECC. Look at the rightmost column, which represents time taken by the
+classical algorithm, and compare it to the "time" columns, which represent how
+much a quantum computer would take. As *n* increases, the amount of time the
+quantum computer would take stays in the same ballpark, whereas for a
```
Comment (catlook) [inline _posts/2016-05-12-supersingular-isogeny-diffie-hellman-101.md]: ... whereas, for a ...

File: _posts/2016-05-12-supersingular-isogeny-diffie-hellman-101.md [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+**What makes this paper special?**
+
+Post-quantum cryptography has been firmly in the realm of academic research
+and experiments. This paper makes significant advancements in how practically
+applicable SIDH is.
+
+**Being future-proof sounds good. If this makes it practical, why don't we
+start using it right now?**
+
+SIDH is a young cryptosystem in a young field, and hasn't had the same level
```
Comment (catlook) [inline _posts/2016-05-12-supersingular-isogeny-diffie-hellman-101.md]: ... field and hasn't ...

File: _posts/2016-05-12-supersingular-isogeny-diffie-hellman-101.md [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+of scrutiny as some of the other post-quantum cryptosystems, let alone the
+"regular" cryptosystems we use daily. Attacks only get better, they never get
```
Comment (catlook) [inline _posts/2016-05-12-supersingular-isogeny-diffie-hellman-101.md]: ... better - they  ...

File: _posts/2016-05-12-supersingular-isogeny-diffie-hellman-101.md [hunk 8]
```diff
@@ -0,0 +1,310 @@
...
+worse. It's possible that SIDH is insecure and we just don't know how to break
```
Comment (catlook) [inline _posts/2016-05-12-supersingular-isogeny-diffie-hellman-101.md]: ... insecure, and we just ...

File: _posts/2016-05-12-supersingular-isogeny-diffie-hellman-101.md [hunk 9]
```diff
@@ -0,0 +1,310 @@
...
+it yet. It does have a good argument for why quantum algorithms wouldn't be
+able to crack it (more on that later), but that's a hypothesis, not a proof.
```
Comment (catlook) [inline _posts/2016-05-12-supersingular-isogeny-diffie-hellman-101.md]: ... hypothesis not a ...

File: _posts/2016-05-12-supersingular-isogeny-diffie-hellman-101.md [hunk 10]
```diff
@@ -0,0 +1,310 @@
...
+SIDH is different. Secret keys are isogenies...
+
+**Whoa whoa whoa. What the heck are isogenies?**
+
+An isogeny between elliptic curves is a function from one elliptic curve to
+another that preserves base points. That means it takes points on one curve
+and returns points on the other curve. Every point on the input curve will map
+to a point on the output curve; but multiple points may map to the same
+point. Formally speaking, the isogeny is surjective. An isogeny is also a
+homomorphism, that is, it preserves the structure of the curve. For any two
```
Comment (catlook) [inline _posts/2016-05-12-supersingular-isogeny-diffie-hellman-101.md]: ... homomorphism. That is, it ...

**Comment by smashwilson** [comment]
Ah: it looks like this would be the first "cryptography" post.

```
Category 'cryptography' only has 1 post
```

Our category linter requires a minimum percentage (2%, currently 7) of posts in the blog to have a category before it's considered valid (to curb the proliferation of loads of categories with only a single post each). To keep it happy you can either:

* Take the "cryptography" category off (which is certainly easier)
* Add "cryptography" to six other existing posts.

File: _posts/2016-05-12-supersingular-isogeny-diffie-hellman-101.md [hunk 11]
```diff
@@ -0,0 +1,310 @@
...
+remember from high school math; it's a single number that tells you something
+interesting about the underlying curve. There are different formulas for
+curves in different forms. For example, for a curve in short Weierstrass form
+`y^2 = x^3 + ax + b`, the j-invariant is:
+
+```
+j(E) = (1728 * 4a^3)/(4a^3 + 27b^2)
+```
+
+The j-invariant has a few cool properties: for example, while this is the
```
Comment (catlook) [inline _posts/2016-05-12-supersingular-isogeny-diffie-hellman-101.md]: ... properties. For  ...

File: _posts/2016-05-12-supersingular-isogeny-diffie-hellman-101.md [hunk 12]
```diff
@@ -0,0 +1,310 @@
...
+formula for the short Weierstrass form, the value of j doesn't change if you
+put the same curve in a different form. Also, all curves with the same
+j-invariant are isomorphic. However, for SIDH you don't really care about
+these properties; you just care that the j-invariant is a number you can
+compute and it'll be the same for the two secret curves that are generated by
```
Comment (catlook) [inline _posts/2016-05-12-supersingular-isogeny-diffie-hellman-101.md]: ... compute, and it'll ...

File: _posts/2016-05-12-supersingular-isogeny-diffie-hellman-101.md [hunk 13]
```diff
@@ -0,0 +1,310 @@
...
+of those two integers with P_A and Q_A to produce a random point R_A, so:
+
+```
+R_A = n_A * P_A + m_A * Q_A
+```
+
+That random point defines Alice's secret isogeny through the isogeny formulas
+I talked about above. The codomain of that isogeny forms Alice's public
+curve. Alice transforms points P_B and Q_B with the isogeny. She sends Bob her
+public curve, and the two transformed points.
```
Comment (catlook) [inline _posts/2016-05-12-supersingular-isogeny-diffie-hellman-101.md]: ... curve and the ...

**Comment by lvh** [comment]
@smashwilson Hey! I can understand why you'd want to have some consistency in categories (e.g. not "crypto", "cryptography", "encryption", and instead just "cryptography"), but does this mean that there's no way to categorize cryptography posts until there's a bunch of them? That seems counterproductive. If you're sure, I'll remove the category.

**Comment by lvh** [comment]
(Also, does that apply to categories that wane in popularity? If it's 2%, then it would be harder to get a new category as time passes, right?)

**Comment by smashwilson** [comment]
Yeah, that's how we've been managing them. They used to be a complete mess until @ktbartholomew got fed up and spent a few days painstakingly cleaning them up. He then wrote [the audit script](https://github.com/ktbartholomew/jekyll-audit) that we've used ever since to *keep* them cleaned.

> (Also, does that apply to categories that wane in popularity? If it's 2%, then it would be harder to get a new category as time passes, right?)

Haha, it sure does. We've had at least one instance where a PR caused the threshold to be rounded up, which made the build complain about a bunch of existing categories...

We could also stand to have the list of existing categories exposed somewhere.

Making it a flat threshold rather than a percentage would certainly make sense to me. I'll let Keith weigh in on that, though, he's the one who did all the legwork on our category audits.

**Comment by ktbartholomew** [comment]
> there's no way to categorize cryptography posts until there's a bunch of them?

Yep. If you're willing to commit to writing enough posts to meet the quota, I'm willing to add a category for it. Without this limitation, we'd be right back where we started with dozens of mostly-unused categories in no time.

> Making it a flat threshold rather than a percentage would certainly make sense to me. I'll let Keith weigh in on that, though, he's the one who did all the legwork on our category audits.

👍  https://github.com/ktbartholomew/jekyll-audit/issues/2

**Comment by lvh** [comment]
I have removed the category. I won't be stopping with this post, but I'll add the category to my posts when/if we get to that number.

**Comment by rackernexus** [comment]
If you want to double-check your formatting, check [the preview](https://staging.developer.rackspace.com/build-2536dc3ebf/).

**Comment by rackernexus** [comment]
:mag_right: [Preview](https://staging.developer.rackspace.com/build-2c8750e30a/)

**Comment by rackernexus** [comment]
:eyes: [Preview](https://staging.developer.rackspace.com/build-5efd9f676c/)

**Comment by lvh** [comment]
@smashwilson, @catlook I've removed the category and made the edits that didn't change the flow away from what I wanted it to be. Thanks a lot for your feedback! I think this is good to go now.

**Comment by smashwilson** [comment]
:+1: :metal: :shipit:

**Pull request merged**